### PR TITLE
feat: remote build command

### PR DESCRIPTION
**Pull request by lengau**
This is a rebase of https://github.com/canonical/charmcraft/pull/1531 on the craft-application-on-main PR

Broken bundle build is fixed in https://github.com/canonical/charmcraft/pull/1550

CRAFT-2544

### CODE REVIEW

**Comment by lengau** [comment]
@sergiusens for this we'll need a `LAUNCHPAD_TOKEN` secret. I believe I've made it replicate snapcraft entirely

File: charmcraft/application/commands/__init__.py
Comment (syu-w) [inline charmcraft/application/commands/__init__.py:121]: Should this count as Lifecycle?

File: charmcraft/application/commands/remote.py
Comment (syu-w) [inline charmcraft/application/commands/remote.py:98]: Why sleep here

File: charmcraft/application/commands/remote.py
Comment (syu-w) [inline charmcraft/application/commands/remote.py:140]: If user Ctrl+C multiple times, this going to skip. It shouldn't be a problem.

**Review by syu-w — APPROVED** [review]

File: charmcraft/application/commands/remote.py
Comment (lengau) [reply charmcraft/application/commands/remote.py:140]: Yep, that's intentional. If someone uses Ctrl+C multiple times they **really** want to quit and I don't want to get in their way. (Especially since the time between a keyboard interrupt and getting that confirmation prompt should be unnoticeable.)

**Review by mr-cal — COMMENTED** [review]
The code looks excellent but I have the following feedback:

1. I would like launchpad credentials to be added to github before this lands, so I can see the spread test pass.
2. Something to do with craft-application is still failing for me :sob: 

```bash
~/dev/charms/simple-charm-from-init> charmcraft remote-build
charmcraft internal error: TypeError("Charmcraft._extra_yaml_transform() missing 2 required keyword-only arguments: 'build_on' and 'build_for'")
Full execution log: '/home/developer/.local/state/charmcraft/log/charmcraft-20240227-125134.189556.log'
```

<details>
  <summary>full logs</summary>

```bash
2024-02-27 12:51:34.197 Starting charmcraft, version 2.5.5.post103+g17a15974
2024-02-27 12:51:34.197 Configuring application...
2024-02-27 12:51:34.197 Preparing application...
2024-02-27 12:51:34.198 Build plan: platform=None, build_for=None
2024-02-27 12:51:34.198 Loading project file '/home/developer/dev/charms/simple-charm-from-init/charmcraft.yaml'
2024-02-27 12:51:34.201 Setting target machine to x86_64
2024-02-27 12:51:34.201 charmcraft internal error: TypeError("Charmcraft._extra_yaml_transform() missing 2 required keyword-only arguments: 'build_on' and 'build_for'")
2024-02-27 12:51:34.202 Traceback (most recent call last):
2024-02-27 12:51:34.202   File "/home/developer/dev/charms/zoneminder-charm/venv/lib/python3.11/site-packages/craft_application/application.py", line 478, in run
2024-02-27 12:51:34.202     self._try_load_project(command, platform, build_for)
2024-02-27 12:51:34.202   File "/home/developer/dev/charms/zoneminder-charm/venv/lib/python3.11/site-packages/craft_application/application.py", line 446, in _try_load_project
2024-02-27 12:51:34.202     self.services.project = self.get_project(
2024-02-27 12:51:34.202                             ^^^^^^^^^^^^^^^^^
2024-02-27 12:51:34.202   File "/home/developer/dev/charms/zoneminder-charm/venv/lib/python3.11/site-packages/craft_application/application.py", line 280, in get_project
2024-02-27 12:51:34.202     yaml_data = self._transform_project_yaml(yaml_data, build_on, build_for)
2024-02-27 12:51:34.202                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-02-27 12:51:34.202   File "/home/developer/dev/charms/zoneminder-charm/venv/lib/python3.11/site-packages/craft_application/application.py", line 570, in _transform_project_yaml
2024-02-27 12:51:34.202     return self._extra_yaml_transform(yaml_data)
2024-02-27 12:51:34.202            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-02-27 12:51:34.202 TypeError: Charmcraft._extra_yaml_transform() missing 2 required keyword-only arguments: 'build_on' and 'build_for'
2024-02-27 12:51:34.202 Full execution log: '/home/developer/.local/state/charmcraft/log/charmcraft-20240227-125134.189556.log'
```

</details>

File: charmcraft/application/commands/remote.py
Comment (lengau) [reply charmcraft/application/commands/remote.py:98]: To give a bit of a pause before moving on. I suppose we could do the same "confirm with user" dance with the Launchpad uploads as well and offer an obnoxiously long CLI parameter so people won't just add it every time they run. @mr-cal what do you think?

File: charmcraft/application/commands/__init__.py
Comment (lengau) [reply charmcraft/application/commands/__init__.py:121]: It looks like that's where Snapcraft has it, so I've adjusted.

**Comment by mr-cal** [comment]
I re-installed craft-application but perhaps it used a cached copy, let me try again

**Review by mr-cal — APPROVED** [review]
The simple case `charmcraft remote-build` works, which is what we discussed was the acceptance criteria of this ticket on 2024-Feb-23, so you have my approval.

The code here is good, but I do have notes:
1. We need credentials for the remote-build spread test
2. riscv doesn't build because charmcraft isn't available on `stable`, which means `charmcraft remote-build` will fail
3. Removing the build-id from `charmcraft remote-build --recover <build-id>` has some unintended consequences and bugs and breaks concurrent builds.
    - For example, I run remote-build, ctrl+c, modify charmcraft.yaml, and can no longer recover the build, because the build-id has changed
    - In another example, I run remote-build, ctrl+c, modify charmcraft.yaml, and cannot run remote-build until the previous build finishes.

**Comment by syu-w** [comment]
> 3. Removing the build-id from `charmcraft remote-build --recover <build-id>` has some unintended consequences and bugs and breaks concurrent builds.
>    
>    * For example, I run remote-build, ctrl+c, modify charmcraft.yaml, and can no longer recover the build, because the build-id has changed
>    * In another example, I run remote-build, ctrl+c, modify charmcraft.yaml, and cannot run remote-build until the previous build finishes.

Seems the build hash is not account for some change?

**Comment by mr-cal** [comment]
> > 3. Removing the build-id from `charmcraft remote-build --recover <build-id>` has some unintended consequences and bugs and breaks concurrent builds.
> >    
> >    * For example, I run remote-build, ctrl+c, modify charmcraft.yaml, and can no longer recover the build, because the build-id has changed
> >    * In another example, I run remote-build, ctrl+c, modify charmcraft.yaml, and cannot run remote-build until the previous build finishes.
> 
> Seems the build hash is not account for some change?

Yeah, I think that is the problem in itself - the build-id is generated from the contents of the files in the project directory to support concurrent builds.  Since `--recover <build-id>` is gone, I think we need to stop doing this and use the path to the project directory as the unique identifier.

**Comment by lengau** [comment]
>     1. We need credentials for the remote-build spread test

I've got a note to ask tomorrow at stand-up, but if for some reason I don't please feel free to interject

>     2. riscv doesn't build because charmcraft isn't available on `stable`, which means `charmcraft remote-build` will fail

Yeah this is an unfortunate thing right now, because the default "all architectures" includes riscv64, so if you want to build your charm not for riscv64 you have to specify architectures in the charm. On the plus side, this is only going into charmcraft 3, which will have a riscv64 build

>     3. Removing the build-id from `charmcraft remote-build --recover <build-id>` has some unintended consequences and bugs and breaks concurrent builds.

Your notes on this have had me generate 3 more work items:

* https://github.com/canonical/craft-application/issues/246
* https://github.com/canonical/craft-application/issues/248
* https://github.com/canonical/craft-application/issues/249

Thank you for your in-depth and thoughtful feedback!

File: charmcraft/application/commands/remote.py
Comment (mr-cal) [reply charmcraft/application/commands/remote.py:98]: I imagine the UX is a little nicer with this pause and 1 second is small compared to the overall time for a remote build, so I'm fine with this.

A comment explaining the delay is for UX would be nice.

Comment (lengau) [reply charmcraft/application/commands/remote.py:99]: Comment added!

**Pull request merged**
